### PR TITLE
Enhancement: Add substitutions for ingame day, month, and year

### PR DESCRIPTION
**Pull request by kydot**
**Feature:** This PR implements the feature request detailed and discussed in issue #9454

## Feature Details
I implemented the required substitutions. For date formats that didn't already exist in Date (i.e. all except for current full date) I created a new function in Date.

## UI Screenshots
N/A

## Usage Examples
This is used the same way that substitutions for first and last name are used. When writing conversations, the text "<current year>" will be substituted with the current year (e.g. "3014"). The other substitutions described in the issue work in the same way.

## Testing Done
Limited testing was done, I just ran the functions and checked that they produced correct output.

### Automated Tests Added
No automated tests were added but if necessary they can be.

## Performance Impact
N/A

### CODE REVIEW

**Review by kydot — COMMENTED** [review]
ConversationPanel: add the substitutions
Date: add the functions used to generate the substitution text
TextReplacements: add the replacement strings to the reserved string list

**Comment by bene-dictator** [comment]
Don't worry about the Ubuntu test failing, it just does that sometimes I'm told.

**Review by OcelotWalrus — APPROVED** [review]
This looks very good to me and I think it has his place in the vanilla game.

File: source/ConversationPanel.cpp [hunk 1]
```diff
@@ -77,6 +78,15 @@ ConversationPanel::ConversationPanel(PlayerInfo &player, const Conversation &con
 	else if(player.Flagship())
 		subs["<ship>"] = player.Flagship()->Name();
 
+	Date date = player.GetDate();
+	subs["<current year>"] = date.CurrentYear();
```
Comment (TomGoodIdea) [inline source/ConversationPanel.cpp:82]:
```suggestion
	subs["<current year>"] = to_string(date.Year());
```
All this function does is taking the output of a single public Date member function and processing it with to_string. This doesn't need to be a separate function IMO.

File: source/Date.cpp [hunk 2]
```diff
@@ -143,7 +143,82 @@ string Date::LongString() const
...
+	int month = Month();
+	int year = Year();
+	string result = Weekday(day, month, year);
+	result += " ";
+	result += std::to_string(day);
+	result += " ";
+	static const string MONTH[] = {
+			"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"};
+	result += MONTH[month - 1];
+	return result;
```
Comment (TomGoodIdea) [inline source/Date.cpp:163]:
```suggestion
	static const string MONTH[] = {
			"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"};
	return Weekday(day, month, Year()) + " " + to_string(day) + " " + MONTH[month - 1];
```
In this case, condensed code is still easy to read + Year() is called once, so you don't need a variable for it.

File: source/Date.cpp [hunk 3]
```diff
@@ -143,7 +143,82 @@ string Date::LongString() const
...
+		dayString += "rd";
+	return dayString;
+}
+string Date::CurrentDayName() const
+{
+	int day = Day();
+	int month = Month();
+	int year = Year();
+	string result = Weekday(day, month, year);
+	return result;
```
Comment (TomGoodIdea) [inline source/Date.cpp:187]:
```suggestion
	return Weekday(Day(), Month(), Year());
```

File: source/Date.cpp [hunk 4]
```diff
@@ -143,7 +143,82 @@ string Date::LongString() const
...
+		"June",
+		"July",
+		"August",
+		"September",
+		"October",
+		"November",
+		"December"
+	};
+	const string &month = MONTH[Month() - 1];
+	return month;
```
Comment (TomGoodIdea) [inline source/Date.cpp:220]:
```suggestion
	return MONTH[Month() - 1];
```

**Review by TomGoodIdea — COMMENTED** [review]
Thank you for working on this!

A few notes:
Please try to maintain the same order of functions in the source file and its header. It makes navigation easier.
Functions in the .cpp file should be separated with triple blank line blocks.
There's a bit of code shared by LongString, CurrentDay and CurrentMonth, responsible for adding suffixes to numbers. I think the Format class is a perfect place for this code.

**Comment by bene-dictator** [comment]
Hi @kydot, are you planning to get to this any time soon?

**Comment by bene-dictator** [comment]
Closing as this hasn't had any activity since last year, though I'd still really like to see this in-game.

**Comment by OcelotWalrus** [comment]
> Closing as this hasn't had any activity since last year, though I'd still really like to see this in-game.

Maybe someone could fork his word and make an other PR? If it gets merged it'll be co-authored of course.

**Comment by bene-dictator** [comment]
That is the hope of an orphaned PR, yes.

**Pull request closed without merging**